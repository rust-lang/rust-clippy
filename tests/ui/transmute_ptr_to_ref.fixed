#![warn(clippy::transmute_ptr_to_ref)]
#![allow(
    clippy::match_single_binding,
    clippy::unnecessary_cast,
    clippy::missing_transmute_annotations
)]

unsafe fn _ptr_to_ref<T, U>(p: *const T, m: *mut T, o: *const U, om: *mut U) {
    unsafe {
        let _: &T = &*p;
        //~^ transmute_ptr_to_ref
        let _: &T = &*p;

        let _: &mut T = &mut *m;
        //~^ transmute_ptr_to_ref
        let _: &mut T = &mut *m;

        let _: &T = &*m;
        //~^ transmute_ptr_to_ref
        let _: &T = &*m;

        let _: &mut T = &mut *(p as *mut T);
        //~^ transmute_ptr_to_ref
        let _ = &mut *(p as *mut T);

        let _: &T = &*(o as *const T);
        //~^ transmute_ptr_to_ref
        let _: &T = &*(o as *const T);

        let _: &mut T = &mut *(om as *mut T);
        //~^ transmute_ptr_to_ref
        let _: &mut T = &mut *(om as *mut T);

        let _: &T = &*(om as *const T);
        //~^ transmute_ptr_to_ref
        let _: &T = &*(om as *const T);
    }
}

fn _issue1231() {
    struct Foo<'a, T> {
        bar: &'a T,
    }

    let raw = 42 as *const i32;
    let _: &Foo<u8> = unsafe { &*raw.cast::<Foo<_>>() };
    //~^ transmute_ptr_to_ref

    let _: &Foo<&u8> = unsafe { &*raw.cast::<Foo<&_>>() };
    //~^ transmute_ptr_to_ref

    type Bar<'a> = &'a u8;
    let raw = 42 as *const i32;
    unsafe { &*(raw as *const u8) };
    //~^ transmute_ptr_to_ref
}

unsafe fn _issue8924<'a, 'b, 'c>(x: *const &'a u32, y: *const &'b u32) -> &'c &'b u32 {
    unsafe {
        match 0 {
            0 => &*x.cast::<&u32>(),
            //~^ transmute_ptr_to_ref
            1 => &*y.cast::<&u32>(),
            //~^ transmute_ptr_to_ref
            2 => &*x.cast::<&'b u32>(),
            //~^ transmute_ptr_to_ref
            _ => &*y.cast::<&'b u32>(),
            //~^ transmute_ptr_to_ref
        }
    }
}

#[clippy::msrv = "1.38"]
unsafe fn _meets_msrv<'a, 'b, 'c>(x: *const &'a u32) -> &'c &'b u32 {
    unsafe {
        let a = 0u32;
        let a = &a as *const u32;
        let _: &u32 = &*a;
        //~^ transmute_ptr_to_ref
        let _: &u32 = &*a.cast::<u32>();
        //~^ transmute_ptr_to_ref
        match 0 {
            0 => &*x.cast::<&u32>(),
            //~^ transmute_ptr_to_ref
            _ => &*x.cast::<&'b u32>(),
            //~^ transmute_ptr_to_ref
        }
    }
}

#[clippy::msrv = "1.37"]
unsafe fn _under_msrv<'a, 'b, 'c>(x: *const &'a u32) -> &'c &'b u32 {
    unsafe {
        let a = 0u32;
        let a = &a as *const u32;
        let _: &u32 = &*a;
        //~^ transmute_ptr_to_ref
        let _: &u32 = &*(a as *const u32);
        //~^ transmute_ptr_to_ref
        match 0 {
            0 => &*(x as *const () as *const &u32),
            //~^ transmute_ptr_to_ref
            _ => &*(x as *const () as *const &'b u32),
            //~^ transmute_ptr_to_ref
        }
    }
}

unsafe fn _issue1754() {
    let mut_num_func_ptr: *mut std::ffi::c_void = std::ptr::null_mut();

    let num: u32 = 0;

    // If pointer cast is supported by MSRV.
    unsafe {
        // Not recommended.
        (&mut *mut_num_func_ptr.cast::<Box<dyn FnMut(&u32)>>())(&num);
        //~^ transmute_ptr_to_ref

        // Recommended.
        (&mut *mut_num_func_ptr.cast::<Box<dyn FnMut(&u32)>>())(&num);
    }

    // If pointer cast is not supported by MSRV.
    #[clippy::msrv = "1.37"]
    unsafe {
        // Not recommended.
        (&mut *(mut_num_func_ptr as *mut Box<dyn FnMut(&u32)>))(&num);
        //~^ transmute_ptr_to_ref

        // Recommended.
        (&mut *(mut_num_func_ptr as *mut Box<dyn FnMut(&u32)>))(&num);
    };

    let num_func_ptr: *const std::ffi::c_void = std::ptr::null_mut();

    // If pointer cast is supported by MSRV.
    unsafe {
        // Not recommended.
        (&*num_func_ptr.cast::<Box<dyn Fn(&u32)>>())(&num);
        //~^ transmute_ptr_to_ref

        // Recommended.
        (&*num_func_ptr.cast::<Box<dyn Fn(&u32)>>())(&num);
    }

    // If pointer cast is not supported by MSRV.
    #[clippy::msrv = "1.37"]
    unsafe {
        // Not recommended.
        (&*(num_func_ptr as *const Box<dyn Fn(&u32)>))(&num);
        //~^ transmute_ptr_to_ref

        // Recommended.
        (&*(num_func_ptr as *const Box<dyn Fn(&u32)>))(&num);
    };
}

fn main() {}
