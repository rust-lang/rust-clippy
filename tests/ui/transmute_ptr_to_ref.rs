#![warn(clippy::transmute_ptr_to_ref)]
#![allow(
    clippy::match_single_binding,
    clippy::unnecessary_cast,
    clippy::missing_transmute_annotations
)]

unsafe fn _ptr_to_ref<T, U>(p: *const T, m: *mut T, o: *const U, om: *mut U) {
    unsafe {
        let _: &T = std::mem::transmute(p);
        //~^ transmute_ptr_to_ref
        let _: &T = &*p;

        let _: &mut T = std::mem::transmute(m);
        //~^ transmute_ptr_to_ref
        let _: &mut T = &mut *m;

        let _: &T = std::mem::transmute(m);
        //~^ transmute_ptr_to_ref
        let _: &T = &*m;

        let _: &mut T = std::mem::transmute(p as *mut T);
        //~^ transmute_ptr_to_ref
        let _ = &mut *(p as *mut T);

        let _: &T = std::mem::transmute(o);
        //~^ transmute_ptr_to_ref
        let _: &T = &*(o as *const T);

        let _: &mut T = std::mem::transmute(om);
        //~^ transmute_ptr_to_ref
        let _: &mut T = &mut *(om as *mut T);

        let _: &T = std::mem::transmute(om);
        //~^ transmute_ptr_to_ref
        let _: &T = &*(om as *const T);
    }
}

fn _issue1231() {
    struct Foo<'a, T> {
        bar: &'a T,
    }

    let raw = 42 as *const i32;
    let _: &Foo<u8> = unsafe { std::mem::transmute::<_, &Foo<_>>(raw) };
    //~^ transmute_ptr_to_ref

    let _: &Foo<&u8> = unsafe { std::mem::transmute::<_, &Foo<&_>>(raw) };
    //~^ transmute_ptr_to_ref

    type Bar<'a> = &'a u8;
    let raw = 42 as *const i32;
    unsafe { std::mem::transmute::<_, Bar>(raw) };
    //~^ transmute_ptr_to_ref
}

unsafe fn _issue8924<'a, 'b, 'c>(x: *const &'a u32, y: *const &'b u32) -> &'c &'b u32 {
    unsafe {
        match 0 {
            0 => std::mem::transmute(x),
            //~^ transmute_ptr_to_ref
            1 => std::mem::transmute(y),
            //~^ transmute_ptr_to_ref
            2 => std::mem::transmute::<_, &&'b u32>(x),
            //~^ transmute_ptr_to_ref
            _ => std::mem::transmute::<_, &&'b u32>(y),
            //~^ transmute_ptr_to_ref
        }
    }
}

#[clippy::msrv = "1.38"]
unsafe fn _meets_msrv<'a, 'b, 'c>(x: *const &'a u32) -> &'c &'b u32 {
    unsafe {
        let a = 0u32;
        let a = &a as *const u32;
        let _: &u32 = std::mem::transmute(a);
        //~^ transmute_ptr_to_ref
        let _: &u32 = std::mem::transmute::<_, &u32>(a);
        //~^ transmute_ptr_to_ref
        match 0 {
            0 => std::mem::transmute(x),
            //~^ transmute_ptr_to_ref
            _ => std::mem::transmute::<_, &&'b u32>(x),
            //~^ transmute_ptr_to_ref
        }
    }
}

#[clippy::msrv = "1.37"]
unsafe fn _under_msrv<'a, 'b, 'c>(x: *const &'a u32) -> &'c &'b u32 {
    unsafe {
        let a = 0u32;
        let a = &a as *const u32;
        let _: &u32 = std::mem::transmute(a);
        //~^ transmute_ptr_to_ref
        let _: &u32 = std::mem::transmute::<_, &u32>(a);
        //~^ transmute_ptr_to_ref
        match 0 {
            0 => std::mem::transmute(x),
            //~^ transmute_ptr_to_ref
            _ => std::mem::transmute::<_, &&'b u32>(x),
            //~^ transmute_ptr_to_ref
        }
    }
}

unsafe fn _issue1754() {
    let mut_num_func_ptr: *mut std::ffi::c_void = std::ptr::null_mut();

    let num: u32 = 0;

    // If pointer cast is supported by MSRV.
    unsafe {
        // Not recommended.
        std::mem::transmute::<*mut std::ffi::c_void, &mut Box<dyn FnMut(&u32)>>(mut_num_func_ptr)(&num);
        //~^ transmute_ptr_to_ref

        // Recommended.
        (&mut *mut_num_func_ptr.cast::<Box<dyn FnMut(&u32)>>())(&num);
    }

    // If pointer cast is not supported by MSRV.
    #[clippy::msrv = "1.37"]
    unsafe {
        // Not recommended.
        std::mem::transmute::<*mut std::ffi::c_void, &mut Box<dyn FnMut(&u32)>>(mut_num_func_ptr)(&num);
        //~^ transmute_ptr_to_ref

        // Recommended.
        (&mut *(mut_num_func_ptr as *mut Box<dyn FnMut(&u32)>))(&num);
    };

    let num_func_ptr: *const std::ffi::c_void = std::ptr::null_mut();

    // If pointer cast is supported by MSRV.
    unsafe {
        // Not recommended.
        std::mem::transmute::<*const std::ffi::c_void, &Box<dyn Fn(&u32)>>(num_func_ptr)(&num);
        //~^ transmute_ptr_to_ref

        // Recommended.
        (&*num_func_ptr.cast::<Box<dyn Fn(&u32)>>())(&num);
    }

    // If pointer cast is not supported by MSRV.
    #[clippy::msrv = "1.37"]
    unsafe {
        // Not recommended.
        std::mem::transmute::<*const std::ffi::c_void, &Box<dyn Fn(&u32)>>(num_func_ptr)(&num);
        //~^ transmute_ptr_to_ref

        // Recommended.
        (&*(num_func_ptr as *const Box<dyn Fn(&u32)>))(&num);
    };
}

fn main() {}
