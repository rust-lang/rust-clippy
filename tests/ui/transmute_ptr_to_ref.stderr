error: transmute from a pointer type (`*const T`) to a reference type (`&T`)
  --> tests/ui/transmute_ptr_to_ref.rs:10:21
   |
LL |         let _: &T = std::mem::transmute(p);
   |                     ^^^^^^^^^^^^^^^^^^^^^^ help: try: `&*p`
   |
   = note: `-D clippy::transmute-ptr-to-ref` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::transmute_ptr_to_ref)]`

error: transmute from a pointer type (`*mut T`) to a reference type (`&mut T`)
  --> tests/ui/transmute_ptr_to_ref.rs:14:25
   |
LL |         let _: &mut T = std::mem::transmute(m);
   |                         ^^^^^^^^^^^^^^^^^^^^^^ help: try: `&mut *m`

error: transmute from a pointer type (`*mut T`) to a reference type (`&T`)
  --> tests/ui/transmute_ptr_to_ref.rs:18:21
   |
LL |         let _: &T = std::mem::transmute(m);
   |                     ^^^^^^^^^^^^^^^^^^^^^^ help: try: `&*m`

error: transmute from a pointer type (`*mut T`) to a reference type (`&mut T`)
  --> tests/ui/transmute_ptr_to_ref.rs:22:25
   |
LL |         let _: &mut T = std::mem::transmute(p as *mut T);
   |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `&mut *(p as *mut T)`

error: transmute from a pointer type (`*const U`) to a reference type (`&T`)
  --> tests/ui/transmute_ptr_to_ref.rs:26:21
   |
LL |         let _: &T = std::mem::transmute(o);
   |                     ^^^^^^^^^^^^^^^^^^^^^^ help: try: `&*(o as *const T)`

error: transmute from a pointer type (`*mut U`) to a reference type (`&mut T`)
  --> tests/ui/transmute_ptr_to_ref.rs:30:25
   |
LL |         let _: &mut T = std::mem::transmute(om);
   |                         ^^^^^^^^^^^^^^^^^^^^^^^ help: try: `&mut *(om as *mut T)`

error: transmute from a pointer type (`*mut U`) to a reference type (`&T`)
  --> tests/ui/transmute_ptr_to_ref.rs:34:21
   |
LL |         let _: &T = std::mem::transmute(om);
   |                     ^^^^^^^^^^^^^^^^^^^^^^^ help: try: `&*(om as *const T)`

error: transmute from a pointer type (`*const i32`) to a reference type (`&_issue1231::Foo<'_, u8>`)
  --> tests/ui/transmute_ptr_to_ref.rs:46:32
   |
LL |     let _: &Foo<u8> = unsafe { std::mem::transmute::<_, &Foo<_>>(raw) };
   |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `&*raw.cast::<Foo<_>>()`

error: transmute from a pointer type (`*const i32`) to a reference type (`&_issue1231::Foo<'_, &u8>`)
  --> tests/ui/transmute_ptr_to_ref.rs:49:33
   |
LL |     let _: &Foo<&u8> = unsafe { std::mem::transmute::<_, &Foo<&_>>(raw) };
   |                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `&*raw.cast::<Foo<&_>>()`

error: transmute from a pointer type (`*const i32`) to a reference type (`&u8`)
  --> tests/ui/transmute_ptr_to_ref.rs:54:14
   |
LL |     unsafe { std::mem::transmute::<_, Bar>(raw) };
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `&*(raw as *const u8)`

error: transmute from a pointer type (`*const &u32`) to a reference type (`&&u32`)
  --> tests/ui/transmute_ptr_to_ref.rs:61:18
   |
LL |             0 => std::mem::transmute(x),
   |                  ^^^^^^^^^^^^^^^^^^^^^^ help: try: `&*x.cast::<&u32>()`

error: transmute from a pointer type (`*const &u32`) to a reference type (`&&u32`)
  --> tests/ui/transmute_ptr_to_ref.rs:63:18
   |
LL |             1 => std::mem::transmute(y),
   |                  ^^^^^^^^^^^^^^^^^^^^^^ help: try: `&*y.cast::<&u32>()`

error: transmute from a pointer type (`*const &u32`) to a reference type (`&&u32`)
  --> tests/ui/transmute_ptr_to_ref.rs:65:18
   |
LL |             2 => std::mem::transmute::<_, &&'b u32>(x),
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `&*x.cast::<&'b u32>()`

error: transmute from a pointer type (`*const &u32`) to a reference type (`&&u32`)
  --> tests/ui/transmute_ptr_to_ref.rs:67:18
   |
LL |             _ => std::mem::transmute::<_, &&'b u32>(y),
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `&*y.cast::<&'b u32>()`

error: transmute from a pointer type (`*const u32`) to a reference type (`&u32`)
  --> tests/ui/transmute_ptr_to_ref.rs:78:23
   |
LL |         let _: &u32 = std::mem::transmute(a);
   |                       ^^^^^^^^^^^^^^^^^^^^^^ help: try: `&*a`

error: transmute from a pointer type (`*const u32`) to a reference type (`&u32`)
  --> tests/ui/transmute_ptr_to_ref.rs:80:23
   |
LL |         let _: &u32 = std::mem::transmute::<_, &u32>(a);
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `&*a.cast::<u32>()`

error: transmute from a pointer type (`*const &u32`) to a reference type (`&&u32`)
  --> tests/ui/transmute_ptr_to_ref.rs:83:18
   |
LL |             0 => std::mem::transmute(x),
   |                  ^^^^^^^^^^^^^^^^^^^^^^ help: try: `&*x.cast::<&u32>()`

error: transmute from a pointer type (`*const &u32`) to a reference type (`&&u32`)
  --> tests/ui/transmute_ptr_to_ref.rs:85:18
   |
LL |             _ => std::mem::transmute::<_, &&'b u32>(x),
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `&*x.cast::<&'b u32>()`

error: transmute from a pointer type (`*const u32`) to a reference type (`&u32`)
  --> tests/ui/transmute_ptr_to_ref.rs:96:23
   |
LL |         let _: &u32 = std::mem::transmute(a);
   |                       ^^^^^^^^^^^^^^^^^^^^^^ help: try: `&*a`

error: transmute from a pointer type (`*const u32`) to a reference type (`&u32`)
  --> tests/ui/transmute_ptr_to_ref.rs:98:23
   |
LL |         let _: &u32 = std::mem::transmute::<_, &u32>(a);
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `&*(a as *const u32)`

error: transmute from a pointer type (`*const &u32`) to a reference type (`&&u32`)
  --> tests/ui/transmute_ptr_to_ref.rs:101:18
   |
LL |             0 => std::mem::transmute(x),
   |                  ^^^^^^^^^^^^^^^^^^^^^^ help: try: `&*(x as *const () as *const &u32)`

error: transmute from a pointer type (`*const &u32`) to a reference type (`&&u32`)
  --> tests/ui/transmute_ptr_to_ref.rs:103:18
   |
LL |             _ => std::mem::transmute::<_, &&'b u32>(x),
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `&*(x as *const () as *const &'b u32)`

error: transmute from a pointer type (`*mut std::ffi::c_void`) to a reference type (`&mut std::boxed::Box<dyn for<'a> std::ops::FnMut(&'a u32)>`)
  --> tests/ui/transmute_ptr_to_ref.rs:117:9
   |
LL |         std::mem::transmute::<*mut std::ffi::c_void, &mut Box<dyn FnMut(&u32)>>(mut_num_func_ptr)(&num);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `(&mut *mut_num_func_ptr.cast::<Box<dyn FnMut(&u32)>>())`

error: transmute from a pointer type (`*mut std::ffi::c_void`) to a reference type (`&mut std::boxed::Box<dyn for<'a> std::ops::FnMut(&'a u32)>`)
  --> tests/ui/transmute_ptr_to_ref.rs:128:9
   |
LL |         std::mem::transmute::<*mut std::ffi::c_void, &mut Box<dyn FnMut(&u32)>>(mut_num_func_ptr)(&num);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `(&mut *(mut_num_func_ptr as *mut Box<dyn FnMut(&u32)>))`

error: transmute from a pointer type (`*const std::ffi::c_void`) to a reference type (`&std::boxed::Box<dyn for<'a> std::ops::Fn(&'a u32)>`)
  --> tests/ui/transmute_ptr_to_ref.rs:140:9
   |
LL |         std::mem::transmute::<*const std::ffi::c_void, &Box<dyn Fn(&u32)>>(num_func_ptr)(&num);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `(&*num_func_ptr.cast::<Box<dyn Fn(&u32)>>())`

error: transmute from a pointer type (`*const std::ffi::c_void`) to a reference type (`&std::boxed::Box<dyn for<'a> std::ops::Fn(&'a u32)>`)
  --> tests/ui/transmute_ptr_to_ref.rs:151:9
   |
LL |         std::mem::transmute::<*const std::ffi::c_void, &Box<dyn Fn(&u32)>>(num_func_ptr)(&num);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `(&*(num_func_ptr as *const Box<dyn Fn(&u32)>))`

error: aborting due to 26 previous errors

